# Imports

In [101]:
import pandas as pd
import numpy as np
import itertools

# Pandas practice and basics

In [61]:
lst = [1, 2, 3, 4, 5]
df = pd.DataFrame([lst, lst[:3], lst, lst, lst, lst, lst, lst, lst], 
                  columns=['A', 'B', 'C', 'D', 'E'])

print(df['D'])
print(df)

0    4.0
1    NaN
2    4.0
3    4.0
4    4.0
5    4.0
6    4.0
7    4.0
8    4.0
Name: D, dtype: float64
   A  B  C    D    E
0  1  2  3  4.0  5.0
1  1  2  3  NaN  NaN
2  1  2  3  4.0  5.0
3  1  2  3  4.0  5.0
4  1  2  3  4.0  5.0
5  1  2  3  4.0  5.0
6  1  2  3  4.0  5.0
7  1  2  3  4.0  5.0
8  1  2  3  4.0  5.0


In [37]:
lst = [1, 2, 3, 4, 5]
df_test = pd.DataFrame(lst)
print(df_test)

   0
0  1
1  2
2  3
3  4
4  5


In [38]:
# simple array
data = np.array(['g', 'e', 'e', 'k', 's'])

ser = pd.Series(data)
print(ser)

0    g
1    e
2    e
3    k
4    s
dtype: object


In [56]:
# intialise data of lists.
data = [{'Name': ['Tom', 'nick', 'krish', 'jack'],
        'Age': [20, 21, 19, 18]}, 
        {'Name3': ['Tom', 'nick', 'krish', 'jack'],
        'Age': [20, 21, 19, 18]}]

# Create DataFrame
df = pd.DataFrame(data)

# Print the output.
print(df)
print(df.columns)

                       Name               Age                     Name3
0  [Tom, nick, krish, jack]  [20, 21, 19, 18]                       NaN
1                       NaN  [20, 21, 19, 18]  [Tom, nick, krish, jack]
Index(['Name', 'Age', 'Name3'], dtype='object')


In [40]:
#delete datafram permenantly
del df

try:
    print(df)
except:
    print("DataFrame is deleted")

DataFrame is deleted


In [41]:
#read a csv
path = '../../../Data Visualisation/Datasets/intermittent-renewables-production-france.csv'

data = pd.read_csv(path)

In [42]:
data.head()

,Date and Hour,Date,StartHour,EndHour,Source,Production,dayOfYear,dayName,monthName
0,2020-07-22 20:00:00+02:00,2020-07-22,20:00:00,21:00:00,Solar,244.0,204,Wednesday,July
1,2020-07-23 07:00:00+02:00,2020-07-23,07:00:00,08:00:00,Solar,223.0,205,Thursday,July
2,2020-07-23 16:00:00+02:00,2020-07-23,16:00:00,17:00:00,Solar,2517.0,205,Thursday,July
3,2020-07-23 19:00:00+02:00,2020-07-23,19:00:00,20:00:00,Solar,658.0,205,Thursday,July
4,2020-07-23 23:00:00+02:00,2020-07-23,23:00:00,24:00:00,Solar,0.0,205,Thursday,July


In [43]:
data.columns

Index(['Date and Hour', 'Date', 'StartHour', 'EndHour', 'Source', 'Production',
       'dayOfYear', 'dayName', 'monthName'],
      dtype='object')

In [44]:
type(data['Date'])

pandas.core.series.Series

# Google Playstore Dataset analysis

## Load dataset

In [45]:
playstore_data = pd.read_csv('../../../Data Visualisation/Datasets/Google playstore/googleplaystore.csv')

## Understanding data

In [46]:
playstore_data.shape

(10841, 13)

In [47]:
playstore_data.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [48]:
playstore_data.head

<bound method NDFrame.head of                                                      App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE  

In [49]:
playstore_data.describe()


,Rating
count,9367.000000
mean,4.193338
std,0.537431
min,1.000000
25%,4.000000
50%,4.300000
75%,4.500000
max,19.000000


In [50]:
# Check if there are null values
print(playstore_data.isnull().sum())

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64


## Data cleaning

- Here we have a total of 10841 rows with 13 columns.
- There are null values present in columns nearly 4 columns with major null values present in rating column. 
- Either handle the null values or remove the rows or remove the column based on the requirments  
- The 'rating' column has the highest null values either remove the column or hand

*Implemented soln*
- Since this is for practice, I will be removing the rows where null values are present

In [51]:
# Store the data in new variable so we dont modify the original data variable
cleaned_data = playstore_data.dropna()

#Verify the data cleansing
print(cleaned_data.shape)
print(cleaned_data.isnull().sum())

(9360, 13)
App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64


## Data analysis

### Average rating for all the applications

In [52]:
sum_of_ratings = cleaned_data['Rating'].values.sum()
application_count = cleaned_data.shape[0]

print(f"Average ratings: {round(sum_of_ratings/application_count, 3)}")

Average ratings: 4.192


In [53]:
cleaned_data['Rating'].describe(include='all')

count    9360.000000
mean        4.191838
std         0.515263
min         1.000000
25%         4.000000
50%         4.300000
75%         4.500000
max         5.000000
Name: Rating, dtype: float64

### Average rating and application count for all the applications by gener

In [151]:
#Get all possible unique values, reduces complexity for further steps
geners = cleaned_data['Genres'].unique()

#Spliting as the string has ; for seperating multiple geners
array = [gener.split(';') for gener in geners]

# Solution 1: Using itertools library
unique_geners1_secondary = set(itertools.chain.from_iterable(array))

# Solution 2: using list comprehension
unique_geners2_secondary = set([item
                                for sublist in array 
                                for item in sublist
                                ])

#Secondary filter to further simplify geners 
primary_gener_list = [primary_gener 
                for primary_gener in unique_geners1_secondary 
                if '&' not in primary_gener
                ]

secondary_gener_list = [secondary_gener 
                  for secondary_gener in unique_geners1_secondary 
                  if '&' in secondary_gener
                  ]

for secondary_gener in secondary_gener_list: 
    flag = 0
    words = secondary_gener.split('&')
    for word in words:
        if str.strip(word) not in primary_gener_list:
            flag = 1
            break

    if flag:
        primary_gener_list.append(secondary_gener)

rating_by_gener = {}
applications_by_gener = {}

for gener in sorted(primary_gener_list):
    filtered_df = cleaned_data[cleaned_data['Genres'].str.contains(gener, na=False)]
    
    applications_by_gener[gener] = len(filtered_df)
    rating_by_gener[gener] = round(
                                (sum(filtered_df['Rating']) /
                                 applications_by_gener[gener]),
                                 3)

    

custom_df = pd.DataFrame({
    'Gener': list(rating_by_gener.keys()),  # The keys (genre names) go into the 'Gener' column
    'Rating': list(rating_by_gener.values()),  # The values (ratings) go into the 'Rating' column
    'Application Count': list(applications_by_gener.values())  # The application counts go into the 'Application Count' column
})

print(custom_df)

                      Gener  Rating  Application Count
0                    Action   4.294                493
1                 Adventure   4.268                211
2                    Arcade   4.308                223
3              Art & Design   4.367                 64
4           Auto & Vehicles   4.190                 73
5                    Beauty   4.279                 42
6                     Board   4.300                 60
7         Books & Reference   4.344                180
8               Brain Games   4.362                 69
9                  Business   4.121                303
10                     Card   4.102                 48
11                   Casino   4.286                 37
12                   Casual   4.185                262
13                   Comics   4.155                 58
14            Communication   4.159                329
15               Creativity   4.311                 36
16                   Dating   3.971                195
17        

In [137]:
filtered_df = cleaned_data[cleaned_data['Genres'].str.contains('Action', na=False)]

### Apps with rating 4.0 or greater

In [79]:
c=0

for i in cleaned_data['Rating']:
    if i > 4:
        c += 1

print(f"There are {c} apps with rating 4 or above")

There are 6796 apps with rating 4 or above


### Application with highest reviews

In [80]:
max_reviews = max(cleaned_data['Reviews'])

print(f"Maximum reviewsof an app: {max_reviews}")
print(cleaned_data[cleaned_data['Reviews'] == max_reviews].T)



Maximum reviewsof an app: 9992
                                  2989
App             GollerCepte Live Score
Category                        SPORTS
Rating                             4.2
Reviews                           9992
Size                               31M
Installs                    1,000,000+
Type                              Free
Price                                0
Content Rating                Everyone
Genres                          Sports
Last Updated              May 23, 2018
Current Ver                        6.5
Android Ver                 4.1 and up


### Application(s) with no reviews or less than 10 reviews

In [90]:
# Using .loc to safely assign the data type to avoid the warning
cleaned_data.loc[:, 'Reviews'] = cleaned_data['Reviews'].astype(float)

low_reviews = cleaned_data[cleaned_data['Reviews']<11]
print(len(low_reviews))

642
